Process map

jordansread opened the pull request:
I think this is handy, but comes down to whether or not we have a stomach for this kind of thing in the viz.yaml
```yaml
process:
  - 
    id: state_geoms_dataframe
    location: cache/state_geoms_dataframe.rds
    reader: rds
    processor: geom_dataframe
    eval_args:
      class: I("state")
      mouseover: "sprintf('hovertext(\"%s\", evt)', name)"
      mouseout: I("hovertext(\" \")")
```
This helps us add attributes quickly, but also brings in a library for non-standard eval and has this pattern for eval_args that we don't see elsewhere. 

This works w/ `vizmake()`, but I couldn't figure out this `make` error (using the makefile instead of remake)
```
vizlab/make/fetch.make: createBlockMakefile(block='fetch',outfile='vizlab/make/fetch.make')
make: *** [vizlab/make/fetch.make] Error 1

Exited with status 2.
```

This also jumped over #23 since that wasn't done, but what I did here for fetch should be tossed and replaced with the solution for #23. 

Feedback? @aappling-usgs @dblodgett-usgs

Commit:
updating for start on #24

Commit:
Merge branch 'master' of github.com:USGS-VIZLAB/vizstorm into process_map

Commit:
start on #24 for process.geoms

Commit:
typo

dblodgett-usgs commented on scripts/process/geom_dataframe.R:
```diff
@@ -0,0 +1,57 @@
+#' data_in <- data.frame(name = c('california','colorado'))
+#' eval_data.frame(data_in, class=I('state'), mouseover = "sprintf('hovertext(\"%s\", evt)', name)")
+#' }
+eval_data.frame <- function(data_in, ...){
```
Is the . here indicating something about object orientation or did it slip in?

dblodgett-usgs replied:
See my main comment on this PR. This "unpacker" function is neat, but could we just use dplyr verbs and do all this in code instead? I'm open to options if I'm missing a big selling point though.

jordansread replied:
Yeah - I hear you. I don't agree with a few of your points, but I do have reservations about what I am proposing to add to the viz.yaml. I am going to move this discussion over to #24 so we can talk about the use-cases, and not get too hung up on this implementation.

dblodgett-usgs replied:
👂 good.

dblodgett-usgs commented:
I'm not a big fan of adding more to viz.yml. I'd also be concerned that we would quickly find a need to construct the contents of eval_args in code anyways. 

Another thing to consider is that `sf` `data.frames` are supposed to help us use dplyr verbs more easily. This goes away from that practice toward something that is a bit less intuitive (IMHO). I think it might be nice to embrace that direction and keep doing the dplyr mutate/ifelse/etc. chains to build up these svg classes and stuff.

I think the pattern in the helper function is great but am not a big fan of do.call and passing ... around unless it's really needed. It's tricky to debug and not familiar to noobs 🤔 like me...

Not saying don't do it, just stating a downside of a "fancy" pattern like this.

dblodgett-usgs commented on scripts/process/geom_dataframe.R:
```diff
@@ -0,0 +1,57 @@
+  if (!is.null(deps[['clip_box']])){
+    # remove all polygons that are outside the box
+    # trim all polgons that overlap the box
+    message('process.geom_dataframe clipping is not implemented yet')
```
See comments in https://github.com/USGS-VIZLAB/vizstorm/issues/23 for how this would work.

dblodgett-usgs reviewed (commented):
See comments, I'm mixed on using this pattern. I think we will end up needing to create process code to create these things anyways and am leery of adding more content to viz.yml.

Commit:
merge conflict

Merge branch 'master' of github.com:USGS-VIZLAB/vizstorm into process_map

# Conflicts:
#	scripts/fetch/viewbox_limits.R
#	viz.yaml

Commit:
updates for sf classes

Commit:
merge conflict

Merge branch 'master' of github.com:USGS-VIZLAB/vizstorm into process_map

# Conflicts:
#	viz.yaml

Commit:
update to new {variable} pattern

Commit:
fix

aappling-usgs commented on scripts/process/geom_dataframe.R:
```diff
@@ -4,8 +4,8 @@
 #' Presently, `spatial_data` is Spatial{X}DataFrame, but this will be updated to `sf`
-#' if `eval_args` are present, they are used to convert the `data` slot by evaluating
-#' simple functions w/ variables in `data` (skipped if `eval_args` isn't present)
+#' if `attributes` are present, they are used to convert the `data` slot by evaluating
```
not `data` but `spatial_data`, right? and maybe rather than "evaluating simple functions w/ variables" say something about the `{ }` templating syntax

jordansread replied:
thanks - I had partially updated the docs from an old commit, and didn't update when I switched to "{var}"

aappling-usgs commented on scripts/process/geom_dataframe.R:
```diff
@@ -21,38 +21,27 @@ process.geom_dataframe <- function(viz){
-  return(as.data.frame(element_args))
-  
-  
+# does this work for "hovertext('{ID}-{dog}' evt)"
```
good Q. does it?

jordansread replied:
no, not the way the substitution part is written now, but I am looking for a better/simpler way. In the meantime, I put `stopifnot(length(data_col) <= 1)` in this function.

Commit:
update docs

Commit:
updating to `mustache`

Commit:
not used. was throwaway.

Commit:
adding clipping

dblodgett-usgs commented on scripts/process/geom_dataframe.R:
```diff
@@ -0,0 +1,68 @@
+#' (`mustache`) syntax. This operation is skipped if `attributes` isn't present.
+#' if `clip_box` is present as a dependency, it is used to clip the geometries in 
+#' `spatial_data`
+#' See `sub_attribute_texte` function.
```
texte?

dblodgett-usgs commented on scripts/process/geom_dataframe.R:
```diff
@@ -0,0 +1,68 @@
+  if (!is.null(deps[['clip_box']])){
+    # remove all polygons that are outside the box
+    # trim all polgons that overlap the box
+    geom_dataframe <- st_intersection(st_buffer(geom_dataframe, 0), deps[['clip_box']])
```
hah. Just going for the 0 buffer. 👍 For whatever reason maps::map("state") causes the issue here.

dblodgett-usgs reviewed (approved):
Looks good. Nice and tidy! Mustache syntax is great. If there are limits to mustache use, it might be worth noting it in the docs?

aappling-usgs commented on scripts/process/geom_dataframe.R:
```diff
@@ -0,0 +1,68 @@
+    })
+    
+    geom_dataframe[attr_names] <- attrs_subbed
+    # dropping data that was in geom_dataframe but not data_out:
```
`data_out`?

aappling-usgs commented on scripts/process/geom_dataframe.R:
```diff
@@ -0,0 +1,68 @@
+#' @return a character vector of length nrow(data)
+sub_attribute_text <- function(template, data){
+  sapply(seq_len(nrow(data)), FUN = function(j) {
+    whisker::whisker.render(template, data = data[j,])
```
cool, i like this use of a well-established templating system

aappling-usgs commented on viz.yaml:
```diff
@@ -30,7 +30,7 @@ info:
     sf:
       repo: CRAN
-      version: 0.4.3
+      version: 0.5-4
```
should we get `sp` in here too?

jordansread replied:
for this PR, I don't use `sp`, but I think we do need it for #28

aappling-usgs reviewed (approved):
Works for me

Commit:
updates after peer review
